dbms/src/IO/HTTPCommon.cpp

```diff
@@ -39,30 +164,21 @@ void setResponseDefaultHeaders(Poco::Net::HTTPServerResponse & response, unsigne
         response.set("Keep-Alive", "timeout=" + std::to_string(timeout.totalSeconds()));
 }
 
-std::unique_ptr<Poco::Net::HTTPClientSession> makeHTTPSession(const Poco::URI & uri, const ConnectionTimeouts & timeouts)
+HTTPSessionPtr makeHTTPSession(const Poco::URI & uri, const ConnectionTimeouts & timeouts)
 {
-    bool is_ssl = static_cast<bool>(uri.getScheme() == "https");
-    std::unique_ptr<Poco::Net::HTTPClientSession> session;
+    const std::string & host = uri.getHost();
+    UInt16 port = uri.getPort();
+    bool https = isHTTPS(uri);
 
-    if (is_ssl)
-#if USE_POCO_NETSSL
-        session = std::make_unique<Poco::Net::HTTPSClientSession>();
-#else
-        throw Exception("ClickHouse was built without HTTPS support", ErrorCodes::FEATURE_IS_NOT_ENABLED_AT_BUILD_TIME);
-#endif
-    else
-        session = std::make_unique<Poco::Net::HTTPClientSession>();
+    auto session = makeHTTPSessionImpl(host, port, https, false);
+    setTimeouts(*session, timeouts);
+    return session;
+}
 
-    session->setHost(DNSResolver::instance().resolveHost(uri.getHost()).toString());
-    session->setPort(uri.getPort());
 
-#if POCO_CLICKHOUSE_PATCH || POCO_VERSION >= 0x02000000
-    session->setTimeout(timeouts.connection_timeout, timeouts.send_timeout, timeouts.receive_timeout);
-#else
-    session->setTimeout(std::max({timeouts.connection_timeout, timeouts.send_timeout, timeouts.receive_timeout}));
-#endif
-
-    return session;
+PooledHTTPSessionPtr makePooledHTTPSession(const Poco::URI & uri, const ConnectionTimeouts & timeouts, size_t per_endpoint_pool_size)
+{
+    return HTTPSessionPool::instance().getSession(uri, timeouts, per_endpoint_pool_size);
```

Comment:
Note: maybe it will lead to artificial replication delay when pool happened to be full? (Due to background pool task delay after exception). Maybe it will lead to 10 second delays when we suddenly want to download many small parts.